File: skills/agent-patterns/templates/go/routing.go
/*
 * Routing Pattern Implementation for Go
 * Classifying inputs and directing to specialized handlers
 */

package agentpatterns

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strconv"
	"strings"
)

// AnthropicClient represents a client for the Anthropic API
type AnthropicClient struct {
	APIKey     string
	HTTPClient *http.Client
}

// MessageRequest represents a request to the Anthropic API
type MessageRequest struct {
	Model     string        `json:"model"`
	MaxTokens int           `json:"max_tokens"`
	Messages  []MessageItem `json:"messages"`
}

// MessageItem represents a message in the conversation
type MessageItem struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// MessageResponse represents a response from the Anthropic API
type MessageResponse struct {
	Content []ContentBlock `json:"content"`
}

// ContentBlock represents a content block in the response
type ContentBlock struct {
	Type string `json:"type"`
	Text string `json:"text,omitempty"`
}

// CreateMessage sends a message to the Anthropic API
func (c *AnthropicClient) CreateMessage(ctx context.Context, prompt, model string, maxTokens int) (string, error) {
	reqBody := MessageRequest{
		Model:     model,
		MaxTokens: maxTokens,
		Messages: []MessageItem{
			{Role: "user", Content: prompt},
		},
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", "https://api.anthropic.com/v1/messages", bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("x-api-key", c.APIKey)
	req.Header.Set("anthropic-version", "2023-06-01")
	req.Header.Set("content-type", "application/json")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("API error (status %d): %s", resp.StatusCode, string(body))
	}

	var msgResp MessageResponse
	if err := json.NewDecoder(resp.Body).Decode(&msgResp); err != nil {
		return "", fmt.Errorf("failed to decode response: %w", err)
	}

	for _, block := range msgResp.Content {
		if block.Type == "text" {
			return block.Text, nil
		}
	}

	return "", fmt.Errorf("no text content in response")
}

// ClassificationResult represents the result of a classification
type ClassificationResult struct {
	Category   string  `json:"category"`
	Confidence float64 `json:"confidence"`
	Reasoning  string  `json:"reasoning"`
}

// Route defines a route with its handler
type Route[T any] struct {
	Category    string
	Description string
	Handler     func(ctx context.Context, input string) (T, error)
}

// Router classifies inputs and directs them to specialized handlers.
//
// Example:
//
//	router := NewRouter[string](client, "claude-sonnet-4-20250514")
//	router.AddRoute(Route[string]{
//	    Category: "technical",
//	    Description: "Technical issues",
//	    Handler: handleTechnical,
//	})
//	result, classification, err := router.Route(ctx, "My app crashed", 0.7)
type Router[T any] struct {
	client   *AnthropicClient
	model    string
	routes   map[string]Route[T]
	fallback func(ctx context.Context, input string) (T, error)
}

// NewRouter creates a new Router
func NewRouter[T any](client *AnthropicClient, model string) *Router[T] {
	return &Router[T]{
		client: client,
		model:  model,
		routes: make(map[string]Route[T]),
	}
}

// AddRoute adds a route with its handler
func (r *Router[T]) AddRoute(route Route[T]) *Router[T] {
	r.routes[route.Category] = route
	return r
}

// SetFallback sets the fallback handler
func (r *Router[T]) SetFallback(handler func(ctx context.Context, input string) (T, error)) *Router[T] {
	r.fallback = handler
	return r
}

// Route classifies input and routes to appropriate handler
func (r *Router[T]) Route(ctx context.Context, input string, confidenceThreshold float64) (T, *ClassificationResult, error) {
	var zero T

	classification, err := r.Classify(ctx, input)
	if err != nil {
		return zero, nil, fmt.Errorf("classification failed: %w", err)
	}

	if classification.Confidence < confidenceThreshold {
		if r.fallback != nil {
			result, err := r.fallback(ctx, input)
			return result, classification, err
		}
		return zero, classification, fmt.Errorf("low confidence (%.2f) and no fallback handler set", classification.Confidence)
	}

	route, exists := r.routes[classification.Category]
	if !exists {
		if r.fallback != nil {
			result, err := r.fallback(ctx, input)
			return result, classification, err
		}
		return zero, classification, fmt.Errorf("no handler for category: %s", classification.Category)
	}

	result, err := route.Handler(ctx, input)
	return result, classification, err
}

// Classify classifies input into a category
func (r *Router[T]) Classify(ctx context.Context, input string) (*ClassificationResult, error) {
	var categories []string
	for _, route := range r.routes {
		categories = append(categories, fmt.Sprintf("- %s: %s", route.Category, route.Description))
	}

	prompt := fmt.Sprintf(`Classify the following input into one of these categories:
%s

Input: %s

Respond with JSON in this exact format:
{
    "category": "<category_name>",
    "confidence": <0.0-1.0>,
    "reasoning": "<brief explanation>"
}`, strings.Join(categories, "\n"), input)

	response, err := r.client.CreateMessage(ctx, prompt, r.model, 256)
	if err != nil {
		return nil, err
	}

	return parseClassificationJSON(response)
}

func parseClassificationJSON(jsonStr string) (*ClassificationResult, error) {
	result := &ClassificationResult{
		Confidence: 0.5,
	}

	// Extract category
	categoryRe := regexp.MustCompile(`"category"\s*:\s*"([^"]*)"`)
	if match := categoryRe.FindStringSubmatch(jsonStr); len(match) > 1 {
		result.Category = match[1]
	}

	// Extract confidence
	confidenceRe := regexp.MustCompile(`"confidence"\s*:\s*([0-9.]+)`)
	if match := confidenceRe.FindStringSubmatch(jsonStr); len(match) > 1 {
		if conf, err := strconv.ParseFloat(match[1], 64); err == nil {
			result.Confidence = conf
		}
	}

	// Extract reasoning
	reasoningRe := regexp.MustCompile(`"reasoning"\s*:\s*"([^"]*)"`)
	if match := reasoningRe.FindStringSubmatch(jsonStr); len(match) > 1 {
		result.Reasoning = match[1]
	}

	return result, nil
}

// Complexity represents task complexity levels
type Complexity int

const (
	ComplexitySimple Complexity = iota
	ComplexityModerate
	ComplexityComplex
)

func (c Complexity) String() string {
	switch c {
	case ComplexitySimple:
		return "Simple"
	case ComplexityModerate:
		return "Moderate"
	case ComplexityComplex:
		return "Complex"
	default:
		return "Unknown"
	}
}

// ModelRouter routes to appropriate model based on task complexity
type ModelRouter struct {
	client              *AnthropicClient
	classificationModel string
}

// NewModelRouter creates a new ModelRouter
func NewModelRouter(client *AnthropicClient, classificationModel string) *ModelRouter {
	return &ModelRouter{
		client:              client,
		classificationModel: classificationModel,
	}
}

// RouteByComplexity routes to appropriate model based on task complexity
func (r *ModelRouter) RouteByComplexity(ctx context.Context, input string) (string, error) {
	complexity, err := r.AssessComplexity(ctx, input)
	if err != nil {
		return "", err
	}

	var model string
	switch complexity {
	case ComplexitySimple:
		model = "claude-3-haiku-20240307"
	case ComplexityModerate:
		model = "claude-sonnet-4-20250514"
	case ComplexityComplex:
		model = "claude-opus-4-20250514"
	default:
		model = "claude-sonnet-4-20250514"
	}

	return r.client.CreateMessage(ctx, input, model, 4096)
}

// AssessComplexity assesses the complexity of a task
func (r *ModelRouter) AssessComplexity(ctx context.Context, input string) (Complexity, error) {
	prompt := fmt.Sprintf(`Assess the complexity of this task on a scale:
- Simple: Factual lookup, simple formatting, basic questions
- Moderate: Analysis, summarization, code review
- Complex: Multi-step reasoning, creative writing, complex coding

Task: %s

Respond with just one word: Simple, Moderate, or Complex`, input)

	response, err := r.client.CreateMessage(ctx, prompt, r.classificationModel, 10)
	if err != nil {
		return ComplexityModerate, err
	}

	switch strings.ToLower(strings.TrimSpace(response)) {
	case "simple":
		return ComplexitySimple, nil
	case "complex":
		return ComplexityComplex, nil
	default:
		return ComplexityModerate, nil
	}
}

// Example usage
func ExampleCustomerServiceRouting() error {
	apiKey := getEnv("ANTHROPIC_API_KEY", "")
	if apiKey == "" {
		return fmt.Errorf("ANTHROPIC_API_KEY environment variable not set")
	}

	client := &AnthropicClient{
		APIKey:     apiKey,
		HTTPClient: &http.Client{},
	}

	router := NewRouter[string](client, "claude-sonnet-4-20250514")

	// Add routes
	router.AddRoute(Route[string]{
		Category:    "technical",
		Description: "Technical issues, bugs, errors",
		Handler: func(ctx context.Context, input string) (string, error) {
			prompt := fmt.Sprintf("You are a technical support specialist. Help with: %s", input)
			return client.CreateMessage(ctx, prompt, "claude-sonnet-4-20250514", 1024)
		},
	})

	router.AddRoute(Route[string]{
		Category:    "billing",
		Description: "Billing, payments, subscriptions",
		Handler: func(ctx context.Context, input string) (string, error) {
			prompt := fmt.Sprintf("You are a billing support specialist. Help with: %s", input)
			return client.CreateMessage(ctx, prompt, "claude-sonnet-4-20250514", 1024)
		},
	})

	router.AddRoute(Route[string]{
		Category:    "general",
		Description: "General inquiries, information requests",
		Handler: func(ctx context.Context, input string) (string, error) {
			prompt := fmt.Sprintf("Help the user with: %s", input)
			return client.CreateMessage(ctx, prompt, "claude-3-haiku-20240307", 1024)
		},
	})

	// Set fallback
	router.SetFallback(func(ctx context.Context, input string) (string, error) {
		return client.CreateMessage(ctx, input, "claude-sonnet-4-20250514", 1024)
	})

	// Route a request
	ctx := context.Background()
	result, classification, err := router.Route(ctx, "My card was charged twice", 0.7)
	if err != nil {
		return err
	}

	fmt.Printf("Category: %s\n", classification.Category)
	fmt.Printf("Confidence: %.2f\n", classification.Confidence)
	fmt.Printf("Response: %s\n", result)

	return nil
}

// Helper function to get environment variable with default
func getEnv(key, defaultValue string) string {
	// Implementation would use os.Getenv in production
	return defaultValue
}

File: skills/agent-patterns/templates/rust/routing.rs
/*!
 * Routing Pattern Implementation for Rust
 * Classifying inputs and directing to specialized handlers
 */

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;
use std::sync::Arc;

/// Anthropic API client (simplified)
#[derive(Clone)]
pub struct AnthropicClient {
    api_key: String,
    http_client: reqwest::Client,
}

#[derive(Serialize)]
struct MessageRequest {
    model: String,
    max_tokens: u32,
    messages: Vec<MessageContent>,
}

#[derive(Serialize, Deserialize)]
struct MessageContent {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct MessageResponse {
    content: Vec<ContentBlock>,
}

#[derive(Deserialize)]
struct ContentBlock {
    #[serde(rename = "type")]
    content_type: String,
    text: Option<String>,
}

impl AnthropicClient {
    pub fn new(api_key: String) -> Self {
        Self {
            api_key,
            http_client: reqwest::Client::new(),
        }
    }

    pub async fn create_message(&self, prompt: &str, model: &str, max_tokens: u32) -> Result<String> {
        let request = MessageRequest {
            model: model.to_string(),
            max_tokens,
            messages: vec![MessageContent {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
        };

        let response = self
            .http_client
            .post("https://api.anthropic.com/v1/messages")
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("content-type", "application/json")
            .json(&request)
            .send()
            .await?
            .json::<MessageResponse>()
            .await?;

        response
            .content
            .into_iter()
            .find(|c| c.content_type == "text")
            .and_then(|c| c.text)
            .context("No text content in response")
    }
}

/// Classification result
#[derive(Debug, Clone)]
pub struct ClassificationResult {
    pub category: String,
    pub confidence: f64,
    pub reasoning: String,
}

/// Route definition
pub struct Route<T> {
    pub category: String,
    pub description: String,
    pub handler: Box<dyn Fn(String) -> Pin<Box<dyn Future<Output = Result<T>> + Send>> + Send + Sync>,
}

/// Router that classifies inputs and directs them to specialized handlers.
///
/// # Example
/// ```rust
/// let router = Router::new(client, "claude-sonnet-4-20250514".to_string());
/// router.add_route(Route {
///     category: "technical".to_string(),
///     description: "Technical issues".to_string(),
///     handler: Box::new(|input| Box::pin(handle_technical(input))),
/// });
/// let (result, classification) = router.route("My app crashed").await?;
/// ```
pub struct Router<T> {
    client: Arc<AnthropicClient>,
    model: String,
    routes: HashMap<String, Route<T>>,
    fallback: Option<Box<dyn Fn(String) -> Pin<Box<dyn Future<Output = Result<T>> + Send>> + Send + Sync>>,
}

impl<T> Router<T> {
    pub fn new(client: Arc<AnthropicClient>, model: String) -> Self {
        Self {
            client,
            model,
            routes: HashMap::new(),
            fallback: None,
        }
    }

    /// Add a route with its handler
    pub fn add_route(&mut self, route: Route<T>) -> &mut Self {
        self.routes.insert(route.category.clone(), route);
        self
    }

    /// Set fallback handler
    pub fn set_fallback(
        &mut self,
        handler: Box<dyn Fn(String) -> Pin<Box<dyn Future<Output = Result<T>> + Send>> + Send + Sync>,
    ) -> &mut Self {
        self.fallback = Some(handler);
        self
    }

    /// Classify input and route to appropriate handler
    pub async fn route(
        &self,
        input: &str,
        confidence_threshold: f64,
    ) -> Result<(T, ClassificationResult)> {
        let classification = self.classify(input).await?;

        if classification.confidence < confidence_threshold {
            if let Some(ref fallback) = self.fallback {
                let result = fallback(input.to_string()).await?;
                return Ok((result, classification));
            }
            anyhow::bail!(
                "Low confidence ({:.2}) and no fallback handler set",
                classification.confidence
            );
        }

        if let Some(route) = self.routes.get(&classification.category) {
            let result = (route.handler)(input.to_string()).await?;
            Ok((result, classification))
        } else if let Some(ref fallback) = self.fallback {
            let result = fallback(input.to_string()).await?;
            Ok((result, classification))
        } else {
            anyhow::bail!("No handler for category: {}", classification.category)
        }
    }

    /// Classify input into a category
    pub async fn classify(&self, input: &str) -> Result<ClassificationResult> {
        let categories: Vec<String> = self
            .routes
            .values()
            .map(|r| format!("- {}: {}", r.category, r.description))
            .collect();

        let prompt = format!(
            r#"Classify the following input into one of these categories:
{}

Input: {}

Respond with JSON in this exact format:
{{
    "category": "<category_name>",
    "confidence": <0.0-1.0>,
    "reasoning": "<brief explanation>"
}}"#,
            categories.join("\n"),
            input
        );

        let response = self.client.create_message(&prompt, &self.model, 256).await?;
        parse_classification_json(&response)
    }
}

fn parse_classification_json(json: &str) -> Result<ClassificationResult> {
    // Simple regex-based parsing (use serde_json in production)
    let category = extract_json_string(json, "category").unwrap_or_default();
    let confidence = extract_json_number(json, "confidence").unwrap_or(0.5);
    let reasoning = extract_json_string(json, "reasoning").unwrap_or_default();

    Ok(ClassificationResult {
        category,
        confidence,
        reasoning,
    })
}

fn extract_json_string(json: &str, key: &str) -> Option<String> {
    let pattern = format!(r#""{}":\s*"([^"]*)""#, key);
    let re = regex::Regex::new(&pattern).ok()?;
    re.captures(json).map(|c| c[1].to_string())
}

fn extract_json_number(json: &str, key: &str) -> Option<f64> {
    let pattern = format!(r#""{}":\s*([0-9.]+)"#, key);
    let re = regex::Regex::new(&pattern).ok()?;
    re.captures(json)
        .and_then(|c| c[1].parse::<f64>().ok())
}

/// Model-based routing by task complexity
pub struct ModelRouter {
    client: Arc<AnthropicClient>,
    classification_model: String,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Complexity {
    Simple,
    Moderate,
    Complex,
}

impl ModelRouter {
    pub fn new(client: Arc<AnthropicClient>, classification_model: String) -> Self {
        Self {
            client,
            classification_model,
        }
    }

    /// Route to appropriate model based on task complexity
    pub async fn route_by_complexity(&self, input: &str) -> Result<String> {
        let complexity = self.assess_complexity(input).await?;

        let model = match complexity {
            Complexity::Simple => "claude-3-haiku-20240307",
            Complexity::Moderate => "claude-sonnet-4-20250514",
            Complexity::Complex => "claude-opus-4-20250514",
        };

        self.client.create_message(input, model, 4096).await
    }

    async fn assess_complexity(&self, input: &str) -> Result<Complexity> {
        let prompt = format!(
            r#"Assess the complexity of this task on a scale:
- Simple: Factual lookup, simple formatting, basic questions
- Moderate: Analysis, summarization, code review
- Complex: Multi-step reasoning, creative writing, complex coding

Task: {}

Respond with just one word: Simple, Moderate, or Complex"#,
            input
        );

        let response = self
            .client
            .create_message(&prompt, &self.classification_model, 10)
            .await?;

        Ok(match response.to_lowercase().trim() {
            "simple" => Complexity::Simple,
            "complex" => Complexity::Complex,
            _ => Complexity::Moderate,
        })
    }
}

// Example usage
pub async fn example_customer_service() -> Result<()> {
    let api_key = std::env::var("ANTHROPIC_API_KEY")
        .context("ANTHROPIC_API_KEY environment variable not set")?;

    let client = Arc::new(AnthropicClient::new(api_key));
    let mut router: Router<String> = Router::new(client.clone(), "claude-sonnet-4-20250514".to_string());

    // Add routes
    router.add_route(Route {
        category: "technical".to_string(),
        description: "Technical issues, bugs, errors".to_string(),
        handler: Box::new(move |input| {
            let client = client.clone();
            Box::pin(async move {
                client
                    .create_message(
                        &format!(
                            "You are a technical support specialist. Help with: {}",
                            input
                        ),
                        "claude-sonnet-4-20250514",
                        1024,
                    )
                    .await
            })
        }),
    });

    let client2 = Arc::new(AnthropicClient::new(
        std::env::var("ANTHROPIC_API_KEY").unwrap(),
    ));
    router.add_route(Route {
        category: "billing".to_string(),
        description: "Billing, payments, subscriptions".to_string(),
        handler: Box::new(move |input| {
            let client = client2.clone();
            Box::pin(async move {
                client
                    .create_message(
                        &format!(
                            "You are a billing support specialist. Help with: {}",
                            input
                        ),
                        "claude-sonnet-4-20250514",
                        1024,
                    )
                    .await
            })
        }),
    });

    // Route a request
    let (response, classification) = router
        .route("My card was charged twice", 0.7)
        .await?;

    println!("Category: {}", classification.category);
    println!("Confidence: {:.2}", classification.confidence);
    println!("Response: {}", response);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_classification() {
        let json = r#"{"category": "technical", "confidence": 0.85, "reasoning": "test"}"#;
        let result = parse_classification_json(json).unwrap();
        assert_eq!(result.category, "technical");
        assert!((result.confidence - 0.85).abs() < 0.01);
    }

    #[test]
    fn test_complexity_enum() {
        assert_ne!(Complexity::Simple, Complexity::Complex);
    }
}
